### DISCUSSION TITLE
How to describe additional transactions in HID?

### DISCUSSION THREAD

**Discussion by DazzlingOkami** [hathach/tinyusb#2384] [Q&A] [answered]
How to configure HID descriptors to support sending 3 transfers per micro-frame in HS device mode. I try to modify wMaxPacketSize. But it's not work. 

Does TinyuSB have any support for additional transaction?

**Comment by DazzlingOkami** [comment]
My config:
#define CFG_TUD_HID_EP_BUFSIZE 1024

My desc_configuration:
```c
static uint8_t const desc_configuration[] =
{
    // Config number (index+1), interface count, string index, total length, attribute, power in mA
    TUD_CONFIG_DESCRIPTOR(1, ITF_NUM_TOTAL, 0, MAIN_CONFIG_TOTAL_LEN, 0, 100),

    TUD_HID_INOUT_DESCRIPTOR(ITF_NUM_HID, STRID_HID_INTERFACE, HID_ITF_PROTOCOL_NONE, sizeof(desc_hid_report), EPNUM_HID_OUT, EPNUM_HID_IN, 1024, 1),

};
```

Xfer speed about 7.8M/s. Consistent with theory.

I have do some tests:
1.modify DESCRIPTOR
```c
static uint8_t const desc_configuration[] =
{
    // Config number (index+1), interface count, string index, total length, attribute, power in mA
    TUD_CONFIG_DESCRIPTOR(1, ITF_NUM_TOTAL, 0, MAIN_CONFIG_TOTAL_LEN, 0, 100),

    TUD_HID_INOUT_DESCRIPTOR(ITF_NUM_HID, STRID_HID_INTERFACE, HID_ITF_PROTOCOL_NONE, sizeof(desc_hid_report), EPNUM_HID_OUT, EPNUM_HID_IN, 1024 + 0x1000, 1),

};
```

Computer cannot recognize. Prompt for insufficient device resources in WIN10.

2.modify CFG_TUD_HID_EP_BUFSIZE to 8192
and

```c
static uint8_t const desc_configuration[] =
{
    // Config number (index+1), interface count, string index, total length, attribute, power in mA
    TUD_CONFIG_DESCRIPTOR(1, ITF_NUM_TOTAL, 0, MAIN_CONFIG_TOTAL_LEN, 0, 100),

    TUD_HID_INOUT_DESCRIPTOR(ITF_NUM_HID, STRID_HID_INTERFACE, HID_ITF_PROTOCOL_NONE, sizeof(desc_hid_report), EPNUM_HID_OUT, EPNUM_HID_IN, 510+ 0x0800, 1),

};
```
this config can work. xfer speed is 14.8M/s.

I need a xfer speed of at least 20M/s. Means PacketSize=1024 and requires 3 transactions per microframe.

Can you give me some advice.

Thanks!

**Comment by HiFiPhile** [comment, marked as answer]
> I need a xfer speed of at least 20M/s.

For this bandwidth I recommend use bulk transfer. 

> 3 transactions per microframe.

You mean 3 transactions for both out and in, or in only ?

> Computer cannot recognize. Prompt for insufficient device resources in WIN10.

If you use `TUD_HID_INOUT_DESCRIPTOR` it will try to allocate 3 transactions for each direction for a total of 6 transactions. Maybe your root hub doesn't have enough bandwidth to allocate all of them.

![image](https://github.com/hathach/tinyusb/assets/4375114/0c22c699-f790-45a8-8a33-66c7e581663c)

**Reply by DazzlingOkami** [reply]
Thank you.

I know your meaning. I noly use out endpoint for large bandwidth.

I have rewritten a description so that the additional transactions number for each port can be defined separately.
```c
#define TUD_HID_INOUT_DESCRIPTOR2(_itfnum, _stridx, _boot_protocol, _report_desc_len, _epout, _epin, _epoutsize, _epinsize, _epoutat, _epinat, _ep_interval) \
  /* Interface */\
  9, TUSB_DESC_INTERFACE, _itfnum, 0, 2, TUSB_CLASS_HID, (uint8_t)((_boot_protocol) ? (uint8_t)HID_SUBCLASS_BOOT : 0), _boot_protocol, _stridx,\
  /* HID descriptor */\
  9, HID_DESC_TYPE_HID, U16_TO_U8S_LE(0x0111), 0, 1, HID_DESC_TYPE_REPORT, U16_TO_U8S_LE(_report_desc_len),\
  /* Endpoint Out */\
  7, TUSB_DESC_ENDPOINT, _epout, TUSB_XFER_INTERRUPT, U16_TO_U8S_LE((_epoutat << 11) + _epoutsize), _ep_interval, \
  /* Endpoint In */\
  7, TUSB_DESC_ENDPOINT, _epin, TUSB_XFER_INTERRUPT, U16_TO_U8S_LE((_epinat << 11) + _epinsize), 10
```

My doubts seem to have been resolved.

Thank you again.
